docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: example
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
+    "foo.example.com":
+    - url: "/foo/*"
+      service:
+        name: foosvc
+        namespace: default
+        port: 80
+    - url: "/bar/*"
+      service:
+        name: barsvc
+        namespace: default
+        port: 80
+    "bar.example.com"
+    - url: "/foobar/*"
+        service:
+          name: foobarsvc
+          namespace: default
+          port: 80
+status:
+  # include host here when we have DNS figured out.
+  address: 104.101.11.39
+```
+
+Put more rigidly:
+
+```go
+// IngressPoint represents a point for ingress traffic.
+type IngressPoint struct {
+	TypeMeta
+	ObjectMeta
+	Spec IngressPointSpec
+	Status IngressPointStatus
+}
+
+// IngressPointSpec describes the ingressPoint the user wishes to exist.
+type IngressPointSpec struct {
+	Host    string
+	IngressPoint map[string][]Path
+}
+
+// IngressPointStatus describes the current state of an ingressPoint.
+type IngressPointStatus struct {
+	Address string
+}
+
+// ServiceRef is a reference to a single service:port.
+type ServiceRef struct {
+	Name      string
+	Namespace string
+	Port      int64
```

Comment:
I'm somewhat biased on load balancers because of history.  Customers / user
generally expected endpoints to be the target of the load balancer.  There
are some weird cases - for instance, if with load balancing do I have to
add the port to my service in order to talk to an endpoint?  Consider a pod
with two ports - one TLS and one HTTP.  The TLS port is for internal
consumers, the HTTP ports are consumed by the load balancer because TLS is
imposed at the load balancer.  Do I need to have HTTP port on the service?
Seems nice, but not truly required.

Is there anything magic about the service that we really need to connect
to?  Why not just have load balancers coordinate with services via
endpoints only (and keep services just as the thing that generate
endpoints).  That's a looser coupling than ingress -> service.  It
discourages us from depending on the service (just the endpoints) directly,
instead encouraging us to stay separate.  However, it might also encourage
duplication of the same data.

On Mon, Sep 14, 2015 at 1:38 PM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/loadbalancing.md
> <https://github.com/kubernetes/kubernetes/pull/12827#discussion_r39422403>
> :
>
> > +// IngressPointSpec describes the ingressPoint the user wishes to exist.
> > +type IngressPointSpec struct {
> > +	Host    string
> > +	IngressPoint map[string][]Path
> > +}
> > +
> > +// IngressPointStatus describes the current state of an ingressPoint.
> > +type IngressPointStatus struct {
> > +	Address string
> > +}
> > +
> > +// ServiceRef is a reference to a single service:port.
> > +type ServiceRef struct {
> > +	Name      string
> > +	Namespace string
> > +	Port      int64
>
> Can you add a commend indicating which port this is - is this the named
> port on the service (so that you have to lookup the service to know what
> the endpoint port is) or is it the endpoint port in the endpoints list (the
> output)?
>
> It's the service port.
>
> Right, shouldn't this be an IntOrString to support either numeric or named
> ports?
> Also, is Port required if the service only has one port?
>
> no and yes. It's
> https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1234.
>
>
> I think we should be explicit here - load balancers may want to only read
> ingress and endpoints (and skip the service). Also load balancers may have
> limitations on types of ports (doesn't support http with connection
> upgrade) and so we should try to encourage an explicit desire from users.
>
> I think it's way more intuitive to create a service and plug that port
> into the ingress point. One shouldn't need to lookup the endpoints, the
> context of this entire resource is service centric. Feel free to argue,
> these are just my thoughts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12827/files#r39422403>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift